pkg/pgclient/client.go

```diff
@@ -297,3 +309,43 @@ func observeStatementCacheState(conn *pgx.Conn) bool {
 	statementCacheLen.Observe(float64(statementCacheSize))
 	return true
 }
+
+func healthCheckRoutine(ctx context.Context, conn pgxconn.PgxConn) {
+	r := prometheus.DefaultRegisterer
+	if env := os.Getenv("IS_TEST"); env == "true" {
+		r = prometheus.NewRegistry()
+	}
+	dbHealthChecks := prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: util.PromNamespace,
+			Subsystem: "database",
+			Name:      "health_checks_total",
+			Help:      "Total number of database health checks performed.",
+		},
+	)
+	dbHealthErrors := prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: util.PromNamespace,
+			Subsystem: "database",
+			Name:      "health_check_errors_total",
+			Help:      "Total number of database health check errors.",
+		},
+	)
+	r.MustRegister(dbHealthChecks, dbHealthErrors)
+	go func() {
+		check := time.NewTicker(time.Minute)
+		defer check.Stop()
+		connection := health.NewHealthChecker(conn)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-check.C:
+			}
+			dbHealthChecks.Inc()
+			if err := connection(); err != nil {
+				dbHealthErrors.Inc()
+			}
+		}
+	}()
```

Comment:
Can we do that on prometheus scrape instead of using go routine?

Reply:
We can do this, but in the case of busy databases, this can be blocking. Hence, it's better to keep this async.= with the Prometheus scrape.

Reply:
Right, this itself runs every 1 min. The #1193 runs 3 mins. We can add this there and make a lot of things simpler here.

Pasting the code here and removing it, so I can directly use when working on that PR

```golang
func healthCheckRoutine(r prometheus.Registerer, ctx context.Context, conn pgxconn.PgxConn) {
	dbHealthChecks := prometheus.NewCounter(
		prometheus.CounterOpts{
			Namespace: util.PromNamespace,
			Subsystem: "database",
			Name:      "health_checks_total",
			Help:      "Total number of database health checks performed.",
		},
	)
	dbHealthErrors := prometheus.NewCounter(
		prometheus.CounterOpts{
			Namespace: util.PromNamespace,
			Subsystem: "database",
			Name:      "health_check_errors_total",
			Help:      "Total number of database health check errors.",
		},
	)
	r.MustRegister(dbHealthChecks, dbHealthErrors)
	go func() {
		check := time.NewTicker(time.Minute)
		defer check.Stop()
		connection := health.NewHealthChecker(conn)
		for {
			select {
			case <-ctx.Done():
				return
			case <-check.C:
			}
			dbHealthChecks.Inc()
			if err := connection(); err != nil {
				dbHealthErrors.Inc()
			}
		}
	}()
}
```